부산대 BE_오선재 1단계 - 상품 API

5seonjae opened the pull request:
상품 관리 미션 - 1단계 상품 API 구현입니다.

상품 추가, 전체 조회, 개별 조회, 수정, 삭제 기능을 구현했습니다.

예외 처리와 검증 기능은 각각 GlobalExceptionHandler.java 와 ProductValidator.java 파일을 만들어 구현했습니다.

5seonjae committed:
feat: 상품 등록 기능 및 예외 처리 구현, README 문서 작성

5seonjae committed:
feat: 상품 전체 조회 기능 구현, README 문서 작성

5seonjae committed:
feat: 상품 개별 조회 기능 및 예외 처리 구현, README 문서 작성

5seonjae committed:
feat: 상품 수정 기능 및 예외 처리 구현, README 문서 작성

5seonjae committed:
feat: 상품 삭제 기능 및 예외 처리 구현, README 문서 작성

5seonjae committed:
docs: README 정리 및 Step 1 기능 목록 완료 표시

iphwade commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -0,0 +1,24 @@
+@RequestMapping("/api/products")
+public class ProductController {
+
+    private final ProductService productService = new ProductService();
```
스프링 프레임워크에서 개선된 방법으로 상태(ex, Service)들을 초기화할 수 있는데요. 왜 이런 기능들을 Spring에서 제공하는지 같이 학습해보면 도움이 될 것 같아요.

5seonjae replied:
말씀해주신 부분 확인했습니다.
Spring의 의존성 주입 개념을 이해하고, 생성자 주입 방식으로 수정하겠습니다.
추천해주신 개선 방향 감사합니다!

iphwade commented on src/main/java/gift/entity/Product.java:
```diff
@@ -0,0 +1,50 @@
+    private int price;
+    private String imageUrl;
+
+    Product() {}
+
+    Product(String name, int price, String imageUrl) {
+        this.name = name;
+        this.price = price;
+        this.imageUrl = imageUrl;
+    }
```
불필요한 생성자로 보여지는데요. 혹시 구현한 이유가 있으실까요-?

5seonjae replied:
처음에는 별다른 생각 없이 습관적으로 기본 생성자를 작성했는데, 리뷰를 보고 다시 확인해보니 현재 코드에서는 불필요한 것 같네요.
찾아보니 JPA를 사용할 경우에는 프록시 생성을 위해 기본 생성자가 필요하다는 걸 알게 되었고, 이번 케이스에서는 사용되지 않으니 제거하는 게 맞을 것 같습니다. 피드백 감사합니다!

iphwade commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
👍

iphwade commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class ProductRepository {
```
Bean으로 관리하는 것의 이점은 있을까요?

5seonjae replied:
처음엔 간단한 로직만 구현하다 보니 Bean으로 등록할 생각을 미처 못 했던 것 같습니다.

말씀해주신 대로 Spring에서는 Repository도 Bean으로 관리하면서 의존성 주입을 통해 사용하는 것이 더 좋은 구조인 것 같아요. 
다음 커밋에서 `@Repository` 어노테이션을 추가해서 Bean으로 등록하고, 사용하는 쪽도 생성자 주입 방식으로 바꾸겠습니다.

좋은 피드백 감사합니다!

5seonjae replied:
Bean으로 등록하면 다음과 같은 이점이 있다고 생각합니다:
- Spring이 생명 주기를 관리해주기 때문에 코드가 더 깔끔하고 안정적이며,
- 의존성 주입(DI)이 가능해져서 테스트나 확장성 측면에서도 유리하고,
- 기본적으로 싱글톤으로 관리되어 리소스 낭비도 줄일 수 있습니다.

iphwade replied:
네, 이번 미션은 스프링의 기능들을 잘 활용해보자는 취지도 보여져서요-! 
스프링의 여러 기능들을 경험해보는 시간을 가져보는 것을 목적으로 해도 좋을 것 같아요.

iphwade commented on src/main/java/gift/entity/Product.java:
```diff
@@ -0,0 +1,50 @@
+
+public class Product {
+
+    private Long id;
```
id는 Long이고 price는 int군요. Wrapper class를 사용한 이유가 있을까요?

5seonjae replied:
좋은 질문 감사합니다!
말씀해주신 것처럼 현재 미션에서는 DB를 사용하지 않기 때문에, id를 꼭 Long으로 선언해야 할 필요는 사실 없습니다. primitive 타입인 long으로도 충분했을 것 같아요.

제가 Long을 사용한 건, JPA 사용 경험이 있어서 습관적으로 작성한 부분이었는데, 이번 미션처럼 DB 없이 메모리에서 처리할 경우엔 long이 더 적절할 수도 있었겠네요. 다음에는 상황에 맞게 타입을 선택하는 데 좀 더 신경 쓰겠습니다! 감사합니다 😊

iphwade commented on src/main/java/gift/validator/ProductValidator.java:
```diff
@@ -0,0 +1,35 @@
+
+        if (!isValidUrl(product.getImageUrl())) {
+            throw new IllegalArgumentException("유효한 이미지 URL이 아닙니다.");
+        }
```
대부분의 경우 product에 있는 상태(name, imageUrl, price 등)을 getter로 가져와서 유효성을 검사하는군요. 
아래 URL(urlStr)과 같이 유효성 검증해보면 어떨까요?

5seonjae replied:
말씀해주신 대로 product의 필드를 미리 변수에 담아 유효성을 검사하도록 리팩토링해보겠습니다.
더 명확한 의도 전달과 유지보수에도 도움이 될 것 같아요. 감사합니다!

iphwade commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -17,4 +18,11 @@ public ResponseEntity<Map<String, String>> handleIllegalArgument(IllegalArgument
+    public ResponseEntity<Map<String, String>> handleNotFound(NoSuchElementException e) {
+        Map<String, String> error = new HashMap<>();
+        error.put("error", e.getMessage());
+        return new ResponseEntity<>(error, HttpStatus.NOT_FOUND);
```
👍 상태코드에 대해 많이 알고 계시고 적용하셨군요.

iphwade commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -38,4 +39,10 @@ public ResponseEntity<Product> getProduct(@PathVariable Long id) {
     }
 
+    @PutMapping("/{id}")
+    public ResponseEntity<Product> updateProduct(@PathVariable Long id, @RequestBody Product updateData) {
```
Product 클래스를 재사용하셨군요. ID는 path로 받고 계셔서 Product.id 필드가 불필요해질 것 같은데요. 분리할 때와 하지 않았을 때를 고려해봐도 좋을 것 같습니다~ :)

5seonjae replied:
피드백 감사합니다! 말씀해주신 대로 id는 path에서 받고 있기 때문에 Product 객체 내 id 필드는 update 요청에서는 실제로 사용되지 않아 혼동을 줄 수 있겠네요.
업데이트용 DTO를 별도로 만들어서 id는 path로 받고, 바디에는 수정 가능한 필드만 받는 방식으로 리팩토링해보겠습니다. 😊

iphwade replied:
그럼 DTO와 Product 코드를 분리하는 이유는 무엇일까요-?

5seonjae replied:
DTO를 Product와 분리하는 주된 이유는 역할의 명확한 분리와 유지보수의 편의성 때문입니다.

1. 도메인 객체를 보호할 수 있습니다. (Encapsulation)
Product는 DB와 직접 매핑되는 도메인 모델(Entity)로 내부 구현이나 비즈니스 규칙을 포함하고 있을 수 있습니다.
이런 Entity를 그대로 API 요청이나 응답에 사용하는 것은 예기치 않은 엔티티의 변경이나 보안상의 문제가 발생할 수 있기 때문에 분리하는 것이 좋다고 생각합니다.

2. 요구사항에 맞게 유연한 구조를 설계할 수 있습니다.
등록, 수정 등 API 에서 받는 필드는 경우에 따라 다를 수 있습니다.
ex) 등록 - 전체 필드 필요 / 수정 - 일부 필드 필요
그렇기에 DTO를 사용하면 API 목적에 맞는 데이터 구조를 자유롭게 설계할 수 있습니다.

3. 코드 유지보수 및 확장성이 향상됩니다.
Entity와 API 요청 / 응답 DTO 가 분리되어 있다면, 어느 한 쪽의 변경이 다른 쪽에 불필요한 영향을 주지 않도록 할 수 있습니다.

4. DTO를 사용하면 입출력 구조를 명확하게 할 수 있고 검증 기능을 분리할 수 있습니다.
DTO는 @NotNull, @Min과 같은 별도의 입력 검증(Validation)을 설정하기 좋습니다.
반면에 Entity는 검증 로직 보다 비즈니스 로직 수행에 집중할 수 있도록 역할을 나누는 것이 좋습니다.

iphwade reviewed (commented):
안녕하세요. 요구 사항을 잘 구현해주셔서 요구 사항보다는 생각해볼만한 점들을 코멘트했습니다.
코멘트 확인해보시고 답변주세요. 답이 없는 주제들도 있으니 선재님의 생각을 편히 적어주셔도 좋습니다. 고생하셨습니다.

iphwade reviewed (changes requested):
Request changes로 변경합니다~

5seonjae committed:
refactor: apply constructor-based DI to ProductController

5seonjae committed:
refactor: 불필요한 기본 생성자 제거

5seonjae committed:
refactor: 상품 수정 기능 관련 전반적인 로직 리팩토링

- Product 엔티티 대신 ProductUpdateDto 사용
- price 필드 타입을 int → Integer로 변경 (null 대응)
- 검증 로직 및 관련 메서드 전반 수정

5seonjae committed:
refactor: Product의 id와 price 필드를 Wrapper → primitive 타입으로 변경

5seonjae committed:
refactor: ProductRepository를 Spring Bean으로 등록하고 DI 적용

5seonjae committed:
refactor: README.md 예외 처리 설명 변경

iphwade commented on src/main/java/gift/validator/ProductValidator.java:
```diff
@@ -8,15 +9,39 @@
+        if (!isValidUrl(imageUrl)) {
             throw new IllegalArgumentException("유효한 이미지 URL이 아닙니다.");
         }
     }
```
필드에 대한 정보는 Product 객체를 생성할 때 검증해보는 건 어떨까요? 
Product 객체가 생성될 때 기본적인 필드 검증이 된다면 어떤 장점이 있을까요?

5seonjae replied:
처음에는 Entity와 DTO의 유효성 검증을 공통화하려고 Validator 클래스를 따로 만들었습니다.

피드백을 받고 곰곰히 생각해보니, Validator 내부에 구현되어 있는 검증 기능들은 Entity의 생성자 검증 방식과 DTO의 입력 검증 방식으로 치환될 수 있다는 것을 알게 되었습니다.

생성자 검증 방식의 장점으로는 다음과 같다고 생각합니다.

1. 불변성과 무결성 보장
유효하지 않은 Product 객체가 생성되는 것을 원천적으로 차단할 수 있다.

2. 코드 중복 최소화
여러 곳에서 별도로 유효성 검사를 하지 않아도 된다.
ex) 컨트롤러, 서비스에서 중복된 검사를 하지 않아도 됨

3. 응집도 상승
관련된 로직이 함께 유지되어서 응집도(Cohesion)가 높아진다.

4. 불변 객체로 확장하기 유리하다.
생성자에서만 값을 검증하면 이후에는 setter 없이 객체를 불변하게 만들 수 있다.
다시 말해, setter가 필요 없어지는 것이다.

5. 테스트 용이성 향상
객체 생성 자체에 실패하기 때문에 잘못된 입력에 대한 테스트가 명확해진다.

다만 이 방식은 입력 상황마다 검증 조건이 달라질 때 Product가 너무 많은 걸 책임지게 될 수 있고, 테스트에서도 무조건 유효한 객체만 생성해야 해서 유연성이 떨어질 수 있다는 단점이 있는 것 같습니다.

위의 단점들은 DTO를 만들어서 검증을 하고, 복잡한 검증 과정의 경우에는 따로 Validator 클래스를 만들어서 구현을 하는 것으로 해결할 수 있다고 생각합니다.
그래서 저는 Entity는 생성자 검증 방식으로, DTO는 상황별로 나눠서 입력 검증을 하도록 리팩토링 하겠습니다.
이렇게 하면 책임도 명확히 나뉘고 코드도 더 깔끔해지는 것 같습니다.

iphwade commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Product> createProduct(@RequestBody Product product) {
```
Product 클래스를 활용해서 외부 Requset 처리를 하고 있네요. 외부 Request에 상품 타입이 추가되고 DB에 저장되는 Product에는 상품 타입 정보의 저장이 필요없을 때는 어떻게 구현될까요?

5seonjae replied:
좋은 지적 감사합니다. 
수정 기능에서는 DTO를 사용하여 구현을 했는데, 급하게 코드를 작성하다보니 등록 기능에도 DTO를 사용하는 것을 놓친 것 같습니다.
말씀해주신 것처럼 외부 요청 처리에는 DTO를 별도로 만들어 사용하는 게 좋을 것 같습니다. 
ProductCreateRequest 같은 DTO를 정의해서, 불필요한 정보가 도메인에 저장되지 않도록 리팩터링하겠습니다.

"외부 Request에 상품 타입이 추가되고 DB에 저장되는 Product에는 상품 타입 정보의 저장이 필요 없을 때는 어떻게 구현될까요?"

상품을 등록하는 상황을 가정해보면, ProductCreateRequestDto와 같은 DTO를 별도로 만들어 요청에서 필요한 모든 정보를 받고, 이 중 Entity에는 실제로 저장이 필요한 필드만 전달하는 방식으로 처리하는 것이 적절할 것 같습니다.

이렇게 하면 외부 요청에는 상품 타입 정보를 포함할 수 있으면서도, DB에는 해당 정보를 저장하지 않도록 할 수 있습니다.

iphwade commented on src/main/java/gift/validator/ProductValidator.java:
```diff
@@ -0,0 +1,67 @@
+import java.net.URL;
+import java.util.NoSuchElementException;
+
+public class ProductValidator {
```
모두 static 메서드로 이루어져 있군요. static 메서드는 모든 곳에서 접근할 수 있으므로 쉽게 의존성이 생기기 쉬워요. Validator 클래스를 한번 없애보는 쪽으로 도전해보는 건 어떨까요?

5seonjae replied:
피드백 감사합니다!

말씀해주신 것처럼 static 메서드는 어디서든 호출이 가능하다 보니 객체지향적인 설계와는 거리가 있고, 잘못 사용될 여지가 있는 것도 사실입니다. 이에 따라 Validator 클래스를 제거하고 도메인 내부 또는 DTO 내부에서 검증하는 방식으로 리팩터링해보려고 합니다.

Entity의 경우, 생성자에서 검증 로직을 처리하면 불변성과 무결성을 자연스럽게 보장할 수 있고, 객체가 항상 유효한 상태로만 존재하게 할 수 있어 도메인 주도 설계 관점에서도 바람직하다고 판단했습니다.

DTO의 경우, Spring에서 제공하는 @Valid 및 제약 어노테이션을 활용하면 별도 Validator 클래스 없이도 검증이 가능하므로 불필요한 추상화를 줄일 수 있을 것 같습니다.

따라서 Validator 클래스를 제거하는 방향으로 리팩터링을 진행해보겠습니다. 좋은 피드백 감사합니다!

iphwade reviewed (approved):
답변을 확인해보니 깊이 있게 잘 학습해나가고 있으신 것 같아요. 
열린 질문들을 많이 드렸는데 좋은 방향으로 잘 답변해주셨습니다. 

이번 단계는 여기서 마무리 하겠습니다.
코멘트는 생각해볼만한 내용으로 달았습니다. 2단계 구현하시면서 답변달아주세요.

고생하셨습니다. :)
